### ISSUE TITLE
Gold bar with exception in VisualStudioWorkspaceImpl.EnsureEditableDocuments when renaming a type (with rename file)

### ISSUE THREAD

**Issue by mavasani** [dotnet/roslyn#45031]
**Version Used**: Latest 16.7 dogfood build

Also reported [AB#1243307](https://devdiv.visualstudio.com/0bdbc590-a062-4c3f-b0f6-9383f67865ee/_workitems/edit/1243307)

**Steps to Reproduce**:

1. Open Compilers.sln from Roslyn repo
2. Attempt to invoke rename type (with rename file) while solution is still loading

**Call stack**:
```
System.Exception: Unable to check out the files from source control.&#x000D;&#x000A;
   at Microsoft.VisualStudio.LanguageServices.Implementation.ProjectSystem.VisualStudioWorkspaceImpl.EnsureEditableDocuments(IEnumerable`1 documents)&#x000D;&#x000A;
   at Microsoft.VisualStudio.LanguageServices.Implementation.ProjectSystem.VisualStudioWorkspaceImpl.ApplyTextDocumentChange(DocumentId documentId, SourceText newText)&#x000D;&#x000A;
   at Microsoft.VisualStudio.LanguageServices.Implementation.ProjectSystem.VisualStudioWorkspaceImpl.ApplyDocumentTextChanged(DocumentId documentId, SourceText newText)&#x000D;&#x000A;
   at Microsoft.CodeAnalysis.Workspace.ApplyChangedDocument(ProjectChanges projectChanges, DocumentId documentId)&#x000D;&#x000A;
   at Microsoft.CodeAnalysis.Workspace.ApplyProjectChanges(ProjectChanges projectChanges)&#x000D;&#x000A;
   at Microsoft.CodeAnalysis.Workspace.TryApplyChanges(Solution newSolution, IProgressTracker progressTracker)&#x000D;&#x000A;
   at Microsoft.VisualStudio.LanguageServices.Implementation.ProjectSystem.VisualStudioWorkspaceImpl.TryApplyChanges(Solution newSolution, IProgressTracker progressTracker)&#x000D;&#x000A;
   at Microsoft.CodeAnalysis.Workspace.TryApplyChanges(Solution newSolution)&#x000D;&#x000A;   at Microsoft.CodeAnalysis.Editor.Implementation.InlineRename.InlineRenameSession.ApplyRename(Solution newSolution, IWaitContext waitContext)&#x000D;&#x000A;
   at Microsoft.CodeAnalysis.Editor.Implementation.InlineRename.InlineRenameSession.CommitCore(IWaitContext waitContext, Boolean previewChanges)&#x000D;&#x000A;
   at Microsoft.CodeAnalysis.Editor.Implementation.InlineRename.InlineRenameSession.&lt;&gt;c__DisplayClass78_0.&lt;CommitWorker&gt;b__0(IWaitContext waitContext)&#x000D;&#x000A;
   at Microsoft.VisualStudio.LanguageServices.Implementation.Utilities.VisualStudioWaitIndicator.Wait(String title, String message, Boolean allowCancel, Boolean showProgress, Action`1 action)&#x000D;&#x000A;
   at Microsoft.CodeAnalysis.Editor.Implementation.InlineRename.InlineRenameSession.CommitWorker(Boolean previewChanges)&#x000D;&#x000A;
   at Microsoft.CodeAnalysis.Editor.Implementation.InlineRename.RenameCommandHandler.ExecuteCommand(ReturnKeyCommandArgs args, CommandExecutionContext context)&#x000D;&#x000A;
   at Microsoft.VisualStudio.UI.Text.Commanding.Implementation.EditorCommandHandlerService.&lt;&gt;c__DisplayClass14_1`1.&lt;ExecuteStartingAtIndex&gt;b__0()&#x000D;&#x000A;
   at Microsoft.VisualStudio.Text.Utilities.GuardedOperations.CallExtensionPoint(Object errorSource, Action call, Predicate`1 exceptionFilter)&#x000D;&#x000A;
--- End of stack trace from previous location where exception was thrown ---&#x000D;&#x000A;
   at Microsoft.VisualStudio.Telemetry.WindowsErrorReporting.WatsonReport.GetClrWatsonExceptionInfo(Exception exceptionObject)</description>
  </entry>
```

**Comment by ryzngard** [comment]
@genlu  is there a possible race condition in EnsureEditableDocuments?

**Comment by ryzngard** [comment]
@genlu  do you mind taking a look? I'm very bloated with bugs :)

**Comment by genlu** [comment]
I will put it on my bug list, no guarantee when I will get to this though :P

**Comment by tmat** [comment]
@genlu Any progress? I hit this almost every time I rename a type.

**Comment by tmat** [comment]
Hitting this all the time :(

**Comment by tmat** [comment]
Another problem is that rename isn't transactional - when this exception occurs the rename changes are partially applied.

**Comment by 333fred** [comment]
Bump this again, I hit this today.

**Comment by genlu** [comment]
Unfortunately, I can't hit a repro. Looking at the source, it all comes down to the result of this call `IVsQueryEditQuerySave2.QueryEditFiles`. @jasonmalinowski @CyrusNajmabadi Do you guys know what this API is for? And who owns it?
http://sourceroslyn.io/#Microsoft.VisualStudio.LanguageServices/Implementation/ProjectSystem/VisualStudioWorkspaceImpl.cs,1408

**Comment by ryzngard** [comment]
`// TODO: meditate about the flags we can pass to this and decide what is most appropriate for Roslyn`

Did anyone do this meditation yet?

**Comment by ryzngard** [comment]
> The methods of the IVsQueryEditQuerySave2 interface should be called immediately before a document is edited for the first time, and immediately before a document is saved.


https://docs.microsoft.com/en-us/dotnet/api/microsoft.visualstudio.shell.interop.ivsqueryeditquerysave2?view=visualstudiosdk-2019

This is implemented by the source control providers. I wonder if we're failing in cases when we shouldn't because of this part

```
ErrorHandler.Failed(result)
```

If the result failed, theoretically that just means we don't have any information about source control and should assume it's okay to edit? The edit will fail if that assumption is false. Maybe we can just log a fault for that case for information and still attempt to let the edit happen

**Comment by jasonmalinowski** [comment]
So for some source control providers like the TFVC one, the call to QueryEditFiles is going to be potentially removing read-only flags on the files. If that were to fail or be aborted then we won't be able to write to files on disk. So I can't imagine we can _entirely_ ignore the value lest we just crash a bit later on. We'll probably have to reach out to the source control owners for guidance here.

> Did anyone do this meditation yet?

Apparently not. (I think I wrote that comment...)

**Comment by DavidKarlas** [comment]
This is happening with GIT repo for me no TFVC here.

**Comment by DavidKarlas** [comment]
I have 100% simple repro:
1) Download this zip file -> [Issue45031.zip](https://github.com/dotnet/roslyn/files/6034321/Issue45031.zip)
  * It contains .csproj file with multiple targets `<TargetFrameworks>netstandard1.4;net40;net45</TargetFrameworks>`
  * Two files/classes where Class3.cs references to Class1.cs
2) Open .csproj file in VS
3) Open Class1.cs file
4) Inline rename, also check that you have rename filename checkbox checked
5) Boom, it fails to rename

From my debugging... Problem is multi-targeting... In AppyProjectChanges for 1st project file is renamed, in pass for 2nd project `array` passed to `QueryEditFiles` has `array[0]==null` because `CurrentSolution.GetDocument(Class1Id)` returns null... Since CurrentSolution was updated in 1st project... Or something like that...

**Comment by teroneko** [comment]
Finally it gets attention. I am facing this half of a year. I already provided feedback with the feedback mechanism of visual studio. Now I applied it once again but opened a feedback by myself.

For reference, here my feedback application:
https://developercommunity.visualstudio.com/t/Renaming-a-class-or-interface-that-is-us/1389762

Here the project:
[ZippedRenameCrashBecauseIAmFrameworkMultiTargeted.zip](https://github.com/dotnet/roslyn/files/6253041/ZippedRenameCrashBecauseIAmFrameworkMultiTargeted.zip

Here the repro steps:
>Well, you have two projects. Project “Dependent” (in short Project B, targets net5.0) which is dependent of project “RenameCrashBecauseIAmFrameworkMultiTargeted” (in short Project A, targets netstandard2.0, netcore3.1, net5.0).
>
>Let’s assume project B owns class StringList which implements IStringList residing at project A and you rename IStringList from within file IStringList.cs (project A), then the following crash occurs with following message: “An exception of type Exception has been encountered. This may be caused by an extension.”

**Comment by CyrusNajmabadi** [comment]
`EnsureEditableDocuments` seems to be quite strange in nature @jasonmalinowski .  It actually just throws raw System.Exceptions.  This certainly can't be how it's intended to work, can it?

**Comment by jasonmalinowski** [comment]
@CyrusNajmabadi: well, it's not intended to fail, because if it's failing we've got problems. :smile: Sure, we can choose a more precise exception type, but that won't making anything sure. If we have a reliable repro we need to just debug this.

**Comment by DavidKarlas** [comment]
Just want to mention, that repro and reason why its failing is explained in https://github.com/dotnet/roslyn/issues/45031#issuecomment-784889615

**Comment by 333fred** [comment]
@jasonmalinowski you can also check out https://github.com/333fred/roslyn/commit/9592dae4bcbde3d71caf14865f178673258461c5 (current head of `333fred:editorformattingservice-featureslayer`, though that might change shortly when I push more commits) and rename the `IEditorFormattingService` type in `\src\Features\Core\Portable\Formatting\IEditorFormattingService.cs` to `IFormattingInteractionService`. Has reproed for me consistently.

**Comment by DavidKarlas** [comment]
Bump, just ran into this again, reminder there is nice repro at https://github.com/dotnet/roslyn/issues/45031#issuecomment-784889615

**Comment by ryzngard** [comment]
> @jmarolf Did you even read my comment [#45031 (comment)](https://github.com/dotnet/roslyn/issues/45031#issuecomment-784889615)?
> 
> > From my debugging... Problem is multi-targeting... In AppyProjectChanges for 1st project file is renamed, in pass for 2nd project `array` passed to `QueryEditFiles` has `array[0]==null` because `CurrentSolution.GetDocument(Class1Id)` returns null... Since CurrentSolution was updated in 1st project... Or something like that...

I'll take a look. Design here was from our discussion on how this should be presented to a user and moving forward. Specifically that showing a gold bar for cases where a provider we don't control (in this case version control) fails in an unrecoverable way. We don't want to show gold bars in those cases.

**Comment by jmarolf** [comment]
@DavidKarlas sorry if I didn't comment on your analysis. This issue appears to have several different ways it can be caused. Your repo is one and we should fix it. We also have internal reports of this happening in non-multi-targeted projects with the git source control provider.

As Andrew said the design discussion was about if there is a valid situation where this exception can happen that isn't just due to bugs. We concluded there is. We need to both fix the bugs that are causing this and also have something for the valid cases where rename cannot complete because source control has locked the files.

**Comment by ryzngard** [comment]
#57725 is fixing the portion that is failing. We will still need to track updating user experience around source control. Potentially worth breaking into a separate item and duping #56467 with this to be resolved by the PR

**Comment by DavidKarlas** [comment]
@ryzngard Thank you for fixing 99% of appearances of gold bar! One might say this bug can be closed as low impact now :)

**Comment by ryzngard** [comment]
Closing this as fixed by #57725 and started #57746  for the updated user experience
